### PR TITLE
Timestamp Service Docs (Timelock Docs, Part 2, in a sense)

### PR DESCRIPTION
**Pull request by jeremyk-91**
* Added a section on the timestamp service, usage and its guarantees.
* Added a bit of detail on how to choose a number of servers.

This is a docs change. I found this useful to help me clarify my thinking about timelock and what kinds of invariants we can demand from Jepsen, and I think it would be good to get this merged (though not as high of a priority).

Also we have in the docs "AtlasDB relies on three distinct services. They are the Key Value Service, the Lock Service, and the Timestamp Service. Each of the services is described in detail below." with no docs on Timestamp, which is not great.

---
This change is [Reviewable](https://reviewable.io/reviews/palantir/atlasdb/1519)

### CODE REVIEW

File: docs/source/services/timestamp_service/overview.rst [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+========
+
+The Timestamp Service exposes an API that allows for clients to request fresh timestamps, which are required by the
+AtlasDB transaction protocol. These timestamps are *logical timestamps* (that is, they have no relationship to actual
+wall-clock time), and are returned as signed 64-bit integers.
+
+Guarantees
+==========
+
+We guarantee that a request for a fresh timestamp will return a strictly greater timestamp than any other timestamp
```
Comment (rhero) [inline docs/source/services/timestamp_service/overview.rst]: The guarantee here is not clear to me. You say we return "a strictly greater timestamp than any other timestamp that may be observed before the request was initiated", but the "observed before" part is not completely clear. 

The example you give below also gives a negative case and I would like to see an example of how we maintain a guarantee.

File: docs/source/services/timestamp_service/clustering.rst [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+The Timestamp Service can be run in clustered mode. The main benefit of this is that you get high availability;
+timestamps can be issued as long as a majority of nodes in a cluster are available and can communicate. It may
+be worth noting that there is a performance tradeoff involved here, as distributed consensus requires us to communicate
+with the other servers in the cluster quite frequently (and thus becomes more costly as more servers are added).
+
+The cluster will agree on a leader using leader election. Thereafter, the leader will be the only node that actually
+issues timestamps (other nodes will return a 503). The leader also writes an upper bound for the timestamps that
+it has given out to a persistent store; in practice, we write this to a timestamp table in the key-value store.
+
+Note that the leader still needs to talk to a quorum of nodes before entertaining a timestamp request, to check
```
Comment (rhero) [inline docs/source/services/timestamp_service/clustering.rst]: This is assuming a 3 node cluster right? In general we require floor(n/2) round trips, i.e. a round trip to a quorum of nodes (not counting yourself).

I think this note adds a little bit of confusion and would probably remove it.

Comment (jeremyk-91) [inline docs/source/services/timestamp_service/clustering.rst]: Okay. I should have said one/two round trips _to a quorum of nodes_ - will change this.

File: docs/source/services/timestamp_service/overview.rst [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+Contiguous Blocks
+=================
+
+Users are allowed to make requests for contiguous blocks of timestamps using the ``getFreshTimestamps`` endpoint.
+This may be useful if one knows it is the only client and/or wishes to have bespoke complex concurrency semantics.
+
+The endpoint will return a contiguous block of timestamps; this is indicated by a lower and upper bound (inclusive
+on both ends). However, we do not guarantee that the size of the timestamp range will be equal to the number of
+timestamps requested for; the onus to check this is on the client. We do guarantee that the range returned will
```
Comment (jeremyk-91) [inline docs/source/services/timestamp_service/overview.rst]: Ah, it is _at most_. I will clarify this.

File: docs/source/services/timestamp_service/overview.rst (hunk 1 above)
Comment (jeremyk-91) [inline docs/source/services/timestamp_service/overview.rst]: Regarding the guarantee: I've changed it to read "we guarantee that a request for a fresh timestamp will return a strictly greater timestamp than any other timestamp
**already returned** by an earlier request." Hope that's clearer.

I've added an example of a 'bad' sequence of events that we guarantee will not happen, as well.

**Comment by rhero** [comment]
It's weird that we have installation docs for the lock service and clustering docs for the timestamp service because both can be installed and clustered, but I also see that the Lock service docs were already there. Otherwise looks good!

**Pull request merged**